modules/tinymce/src/core/main/ts/api/dom/SelectorChanged.ts

```diff
@@ -16,56 +21,58 @@ const deleteFromCallbackMap = (callbackMap, selector, callback) => {
 
   const findMatchingNode = (selector: string, nodes: Node[]): Optional<Node> =>
     Arr.find(nodes, (node) => dom.is(node, selector));
 
   const getParents = (elem: Element): Element[] =>
-    dom.getParents(elem, null, dom.getRoot());
-
-  return {
-    selectorChangedWithUnbind: (selector: string, callback: SelectorChangedCallback): { unbind: () => void } => {
-      if (!selectorChangedData) {
-        selectorChangedData = {};
-        currentSelectors = {};
-
-        editor.on('NodeChange', (e) => {
-          const node = e.element;
-          const parents = getParents(node);
-          const matchedSelectors = {};
-
-          // Check for new matching selectors
-          Tools.each(selectorChangedData, (callbacks, selector) => {
-            findMatchingNode(selector, parents).each((node) => {
-              if (!currentSelectors[selector]) {
-                // Execute callbacks
-                Arr.each(callbacks, (callback) => {
-                  callback(true, { node, selector, parents });
-                });
-
-                currentSelectors[selector] = callbacks;
-              }
-
-              matchedSelectors[selector] = callbacks;
+    dom.getParents(elem, undefined, dom.getRoot());
+
+  const setup = (): void => {
+    selectorChangedData = {};
+    currentSelectors = {};
+
+    editor.on('NodeChange', (e) => {
+      const node = e.element;
+      const parents = getParents(node);
+      const matchedSelectors: Record<string, SelectorChangedCallback[]> = {};
+
+      // Check for new matching selectors
+      Obj.each(selectorChangedData, (callbacks, selector) => {
+        findMatchingNode(selector, parents).each((node) => {
+          if (!currentSelectors[selector]) {
+            // Execute callbacks
+            Arr.each(callbacks, (callback) => {
+              callback(true, { node, selector, parents });
             });
-          });
 
-          // Check if current selectors still match
-          Tools.each(currentSelectors, (callbacks, selector) => {
-            if (!matchedSelectors[selector]) {
-              delete currentSelectors[selector];
+            currentSelectors[selector] = callbacks;
+          }
 
-              Tools.each(callbacks, (callback) => {
-                callback(false, { node, selector, parents });
-              });
-            }
-          });
+          matchedSelectors[selector] = callbacks;
         });
+      });
+
+      // Check if current selectors still match
+      Obj.each(currentSelectors, (callbacks, selector) => {
+        if (!matchedSelectors[selector]) {
+          delete currentSelectors[selector];
+
+          Arr.each(callbacks, (callback) => {
+            callback(false, { node, selector, parents });
+          });
+        }
+      });
+    });
+  };
+
+  return {
+    selectorChangedWithUnbind: (selector: string, callback: SelectorChangedCallback): { unbind: () => void } => {
+      if (!selectorChangedData) {
+        setup();
```

Comment:
Abstracted the setup logic out into a separate function to improve readability here and help with the types a little.